go/pkg/cas/upload.go

```diff
@@ -662,6 +677,160 @@ func (u *uploader) uploadBatch(ctx context.Context, reqs []*repb.BatchUpdateBlob
 	})
 }
 
+// stream uploads the item using ByteStream service.
+//
+// If the blob is already uploaded, then the function returns quickly and
+// without an error.
+func (u *uploader) stream(ctx context.Context, item *uploadItem, updateCacheStats bool) error {
+	if err := u.semByteStreamWrite.Acquire(ctx, 1); err != nil {
+		return err
+	}
+	defer u.semByteStreamWrite.Release(1)
+
+	// Prepare compressor, if needed.
+	enableCompression := item.Digest.SizeBytes >= u.CompressedBytestreamThreshold
+	var enc *zstd.Encoder
+	if enableCompression {
+		enc = zstdEncoders.Get().(*zstd.Encoder)
+		defer func() {
+			enc.Close()
+			zstdEncoders.Put(enc)
+		}()
+	}
+
+	streamBuf := u.streamBufs.Get().([]byte)
+	defer u.streamBufs.Put(streamBuf)
+
+	// Open the item.
+	r, err := item.Open()
+	if err != nil {
+		return err
+	}
+	defer r.Close()
+
+	// TODO(nodir): implement per-RPC timeouts. No nice way to do it.
+	rewind := false
+	return u.withRetries(ctx, func(ctx context.Context) error {
+		// TODO(nodir): check QueryWriteStatus.
```

Comment:
FWIW, I just sanity checked here - RBE doesn't currently support resumable uploads (ie we never "commit" non-full blobs). In case of eg local timeouts, multiple in flight uploads are valid but only one gets committed. In the current implementation we always reset to zero, and that happens to work.

I'm ok if you want to leave the TODO, just mention this won't work for backends that implement resumable uplodas.

Reply:
> You don't need to check QueryWriteStatus - that'd be an unnecessary RPC call. While not explicitly stated, I can't foresee a sane backend implementation query QueryWriteStatus.CommittedSize is diff from WriteResponse.CommittedSize. You can check the response on stream.CloseAndRecv();.

Sorry, I am not sure I understood this. Do you mean that instead of making `QueryWriteStatus` RPC, we can opportunistically start streaming assuming `CommittedSize=0`, but then check `WriteResponse.CommittedSize`. We don't get `WriteResponse.CommittedSize` until we start streaming, and we need to decide on the offset before we start streaming.

> I also don't think you can leave the withRetries with SeekStart(0) here. That'd make the retries useless (ie they'd never succeed. Either remove the retrying for now or fix the seek.

Could you elaborate on why this wouldn't work? Suppose we started streaming, something went wrong and we closed the stream in defer. I assume the server will throw the results and reset to its original state. On the second attempt we start on a clean slate.

I've changed TODO to "add support for resumable uploads", since it is conceivable to implement it on the server (at least without compression)